contrib/educational_decoder/zstd_decompress.c

```diff
@@ -0,0 +1,2168 @@
+                INP_SIZE();
+            }
+            src += block_len;
+            break;
+        case 3:
+            // Reserved block type
+            CORRUPTION();
+            break;
+        }
+    } while (!last_block);
+
+    if (header->content_checksum_flag) {
+        if (src + 4 > src_end) {
+            INP_SIZE();
+        }
+        src += 4;
+    }
+
+    return src - src_beg;
+}
+
+/******* END OUTPUT SIZE COUNTING *********************************************/
+
+/******* DICTIONARY PARSING ***************************************************/
+static void init_raw_content_dict(dictionary_t *const dict, const u8 *const src,
+                                  const size_t src_len);
+
+static void parse_dictionary(dictionary_t *const dict, const u8 *src,
+                             size_t src_len) {
+    memset(dict, 0, sizeof(dictionary_t));
+    if (src_len < 8) {
+        INP_SIZE();
+    }
+    const u32 magic_number = read_bits_LE(src, 32, 0);
+    if (magic_number != 0xEC30A437) {
+        // raw content dict
+        init_raw_content_dict(dict, src, src_len);
+        return;
+    }
+    dict->dictionary_id = read_bits_LE(src, 32, 32);
+
+    src += 8;
+    src_len -= 8;
+
+    // Parse the provided entropy tables in order
+    {
+        const size_t read =
+                decode_huf_table(src, src_len, &dict->literals_dtable);
+        src += read;
+        src_len -= read;
+    }
+    {
+        const size_t read = decode_seq_table(src, src_len, &dict->of_dtable,
+                                             seq_offset, seq_fse);
+        src += read;
+        src_len -= read;
+    }
+    {
+        const size_t read = decode_seq_table(src, src_len, &dict->ml_dtable,
+                                             seq_match_length, seq_fse);
+        src += read;
+        src_len -= read;
+    }
+    {
+        const size_t read = decode_seq_table(src, src_len, &dict->ll_dtable,
+                                             seq_literal_length, seq_fse);
+        src += read;
+        src_len -= read;
+    }
+
+    if (src_len < 12) {
+        INP_SIZE();
+    }
+    // Read in the previous offset history
+    dict->previous_offsets[1] = read_bits_LE(src, 32, 0);
+    dict->previous_offsets[2] = read_bits_LE(src, 32, 32);
+    dict->previous_offsets[3] = read_bits_LE(src, 32, 64);
+
+    src += 12;
+    src_len -= 12;
+
+    // Ensure the provided offsets aren't too large
+    for (int i = 1; i <= 3; i++) {
+        if (dict->previous_offsets[i] > src_len) {
+            ERROR("Dictionary corrupted");
+        }
+    }
+    // The rest is the content
+    dict->content = malloc(src_len);
+    if (!dict->content) {
+        BAD_ALLOC();
+    }
+
+    dict->content_size = src_len;
+    memcpy(dict->content, src, src_len);
+}
+
+/// If parse_dictionary is given a raw content dictionary, it delegates here
+static void init_raw_content_dict(dictionary_t *const dict, const u8 *const src,
```

Comment:
Change this to `init_dictionary_content()`, that way it can be used both for raw content dictionaries and zstd dictionary content.

```c
if (/* raw content dictionary */) {
  dict->dictionary_id = 0;
  init_dictionary_content();
}
// ...
init_dictionary_content();
return;